File: solutions/JavaScript/generate_parentheses_dp.js
function generateParenthesis(n) {
    const dp = Array(n + 1).fill(null).map(() => []);
    dp[0] = [""];

    for (let i = 1; i <= n; i++) {
        for (let j = 0; j < i; j++) {
            for (const left of dp[j]) {
                for (const right of dp[i - 1 - j]) {
                    dp[i].push("(" + left + ")" + right);
                }
            }
        }
    }

    return dp[n];
}

File: solutions/Rust/generate_parentheses_dp.rs
pub fn generate_parenthesis(n: i32) -> Vec<String> {
    let n = n as usize;
    let mut dp = vec![vec![]; n + 1];
    dp[0] = vec!["".to_string()];

    for i in 1..=n {
        for j in 0..i {
            for left in &dp[j] {
                for right in &dp[i - 1 - j] {
                    dp[i].push(format!("({}){}", left, right));
                }
            }
        }
    }

    dp[n].clone()
}
